File: ios/sdk/swift/PeerNodeListener.swift
//
//  PeerListener.swift
//  sdk
//
//  Created by mengxk on 2019/12/9.
//  Copyright © 2019 Elastos. All rights reserved.
//

import Foundation
import ContactSDK

protocol FileInformationProviderDelegate { }

public protocol PeerNodeListenerListener: class {
    func onAcquire(request: Contact.Listener.AcquireArgs) -> Data?
    
    func onError(errCode: Int,
                 errStr: String,
                 ext: String?)
}


public protocol PeerNodeListenerMessageListener: class {
    func onEvent(event: Contact.Listener.EventArgs)

    func onReceivedMessage(humanCode: String,
                           channelType: Contact.Channel,
                           message: Contact.Message)
}

public protocol PeerNodeListenerDataListener: class {
    func onNotify(humanCode: String,
                  channelType: Contact.Channel,
                  dataId: String,
                  status: Contact.DataListener.Status)

    func onReadData(humanCode: String,
                    channelType: Contact.Channel,
                    dataId: String,
                    offset: Int64,
                    data: inout Data?) -> Int

    func onWriteData(humanCode: String,
                     channelType: Contact.Channel,
                     dataId: String,
                     offset: Int64,
                     data: Data?) -> Int
}

open class PeerNodeListener {
    public typealias Listener = PeerNodeListenerListener
    weak var listener: Listener?

    public typealias MessageListener = PeerNodeListenerMessageListener
    weak var messageListener: MessageListener?

    public typealias DataListener = PeerNodeListenerDataListener
    weak var dataListener: DataListener?
}

File: android/sdk/src/main/java/app/elaphant/sdk/peernode/PeerNodeListener.java
package app.elaphant.sdk.peernode;

import org.elastos.sdk.elephantwallet.contact.Contact;
import org.elastos.sdk.elephantwallet.contact.internal.ContactChannel;
import org.elastos.sdk.elephantwallet.contact.internal.ContactDataListener;
import org.elastos.sdk.elephantwallet.contact.internal.ContactListener;

import java.nio.ByteBuffer;

public class PeerNodeListener {

    public interface Listener {
        byte[] onAcquire(ContactListener.AcquireArgs request);
        void onError(int errCode, String errStr, String ext);
    }

    public interface MessageListener {
        void onEvent(ContactListener.EventArgs event);
        void onReceivedMessage(String humanCode, ContactChannel channelType, Contact.Message message);
    }

    public interface DataListener {
        void onNotify(String humanCode, ContactChannel channelType,
                                      String dataId, ContactDataListener.Status status);
        int onReadData(String humanCode, ContactChannel channelType,
                                       String dataId, long offset, ByteBuffer data);
        int onWriteData(String humanCode, ContactChannel channelType,
                                        String dataId, long offset, byte[] data);
    }


}
